### PR TITLE
Refactor CR delete/update log

### PR DESCRIPTION
**Pull request by yanjunz97**

### CODE REVIEW

File: pkg/controllers/networkpolicy/networkpolicy_controller.go [hunk 1]
```diff
@@ -44,23 +44,30 @@ type NetworkPolicyReconciler struct {
...
+	}
 	r.Recorder.Event(o, v1.EventTypeWarning, common.ReasonFailUpdate, fmt.Sprintf("%v", *e))
 	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, MetricResType)
 }
 
-func deleteFail(r *NetworkPolicyReconciler, c context.Context, o *networkingv1.NetworkPolicy, e *error) {
-	r.Recorder.Event(o, v1.EventTypeWarning, common.ReasonFailDelete, fmt.Sprintf("%v", *e))
+func deleteFail(r *NetworkPolicyReconciler, namespacedName types.NamespacedName, e *error) {
+	log.Error(*e, "Failed to delete NetworkPolicy, would retry exponentially", "NetworkPolicy", namespacedName)
+
```
Comment (zhengxiexie) [inline pkg/controllers/networkpolicy/networkpolicy_controller.go:59]: Remove the blank

Comment (yanjunz97) [reply pkg/controllers/networkpolicy/networkpolicy_controller.go]: Updated, thanks!

**Review by zhengxiexie — DISMISSED** [review]
LGTM

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 2]
```diff
@@ -42,23 +42,25 @@ type IPAddressAllocationReconciler struct {
...
-func deleteSuccess(r *IPAddressAllocationReconciler, _ context.Context, o *v1alpha1.IPAddressAllocation) {
-	r.Recorder.Event(o, v1.EventTypeNormal, common.ReasonSuccessfulDelete, "IPAddressAllocation CR has been successfully deleted")
+func deleteSuccess(r *IPAddressAllocationReconciler, namespacedName types.NamespacedName) {
+	log.Info("Successfully deleted IPAddressAllocation CR", "IPAddressAllocation", namespacedName)
 	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResType)
 }
 
-func deleteFail(r *IPAddressAllocationReconciler, c context.Context, o *v1alpha1.IPAddressAllocation, e *error) {
-	r.Recorder.Event(o, v1.EventTypeWarning, common.ReasonFailDelete, fmt.Sprintf("%v", *e))
+func deleteFail(r *IPAddressAllocationReconciler, namespacedName types.NamespacedName, e *error) {
```
Comment (wenyingd) [inline pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:50]: Why use a pointer for the error in signature? Can we use type `error` directly?

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 3]
```diff
@@ -42,23 +42,25 @@ type IPAddressAllocationReconciler struct {
...
 
 func updateSuccess(r *IPAddressAllocationReconciler, c context.Context, o *v1alpha1.IPAddressAllocation) {
+	log.Info("Successfully created or updated IPAddressAllocation CR", "IPAddressAllocation", *o)
 	r.setReadyStatusTrue(c, o, metav1.Now())
 	r.Recorder.Event(o, v1.EventTypeNormal, common.ReasonSuccessfulUpdate, "IPAddressAllocation CR has been successfully updated")
 	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateSuccessTotal, MetricResType)
 }
 
 func updateFail(r *IPAddressAllocationReconciler, c context.Context, o *v1alpha1.IPAddressAllocation, e *error) {
+	log.Error(*e, "Failed to create or update IPAddressAllocation CR, would retry exponentially", "IPAddressAllocation", *o)
```
Comment (wenyingd) [inline pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go:63]: ditto

File: pkg/controllers/networkinfo/networkinfo_utils.go [hunk 4]
```diff
@@ -19,27 +19,28 @@ import (
 	svccommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
 )
 
-func deleteFail(r *NetworkInfoReconciler, c context.Context, o *v1alpha1.NetworkInfo, e *error, client client.Client) {
-	setNetworkInfoVPCStatus(c, o, client, nil)
-	r.Recorder.Event(o, v1.EventTypeWarning, common.ReasonFailDelete, fmt.Sprintf("%v", *e))
+func deleteFail(r *NetworkInfoReconciler, namespacedName apitypes.NamespacedName, e *error) {
+	log.Error(*e, "Failed to delete stale NSX VPC, would retry exponentially", "Networkinfo", namespacedName)
```
Comment (wenyingd) [inline pkg/controllers/networkinfo/networkinfo_utils.go:23]:
```suggestion
	log.Error(*e, "Failed to delete NSX VPC, would retry exponentially", "Networkinfo", namespacedName)
```

**Review by wenyingd — COMMENTED** [review]
Do you think we can provide a common utility for such functions `deleteFail, updateFail, deleteSuccess`, the fields for the utility could be NSXConfig, and each call can provide apitypes.NamespacedName which presents the K8s object. Then the major difference is the log sentence, maybe we can also make it with a general format, and the caller can provide the Resource name?

File: pkg/controllers/common/utils.go [hunk 5]
```diff
@@ -159,3 +162,12 @@ func GenericGarbageCollector(cancel chan bool, timeout time.Duration, f func(ctx
 		}
 	}
 }
+
+func UpdateSuccess(ctx context.Context, obj k8sclient.Object, resourceType string, metricResType string, NSXConfig *config.NSXOperatorConfig, recorder record.EventRecorder, client k8sclient.Client, setStatusFn func(k8sclient.Client, context.Context, k8sclient.Object, metav1.Time)) {
+	log.Info(fmt.Sprintf("Successfully created or updated %s CR", resourceType), resourceType, obj)
+	if setStatusFn != nil {
+		setStatusFn(client, ctx, obj, metav1.Now())
```
Comment (wenyingd) [inline pkg/controllers/common/utils.go:169]: Use a type definition for func `setStatusFn`? e.g.,

```
type UpdateStatusFn func(k8sclient.Client, context.Context, k8sclient.Object, metav1.Time)

type StatusUpdater struct {
	Client        k8sclient.Client
	NSXConfig     *config.NSXOperatorConfig
	resourceType  string
	recorder      record.EventRecorder
	metricResType string
}

func (u *StatusUpdater) UpdateSuccess(ctx context.Context, obj k8sclient.Object, setStatusFn UpdateStatusFn) {
	log.Info(fmt.Sprintf("Successfully created or updated %s CR", u.resourceType), u.resourceType, obj)
	if setStatusFn != nil {
		setStatusFn(u.Client, ctx, obj, metav1.Now())
	}
	u.recorder.Event(obj, v1.EventTypeNormal, ReasonSuccessfulUpdate, fmt.Sprintf("%s CR has been successfully updated", u.resourceType))
	metrics.CounterInc(u.NSXConfig, metrics.ControllerUpdateSuccessTotal, u.metricResType)
}
```

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go (hunk 2 above)
Comment (yanjunz97) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go]: Updated, thanks

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go (hunk 3 above)
Comment (yanjunz97) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go]: updated, thanks

File: pkg/controllers/subnetset/subnetset_controller.go [hunk 6]
```diff
@@ -55,15 +55,25 @@ func (r *SubnetSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		log.Info("Finished reconciling SubnetSet", "SubnetSet", req.NamespacedName, "duration(ms)", time.Since(startTime).Milliseconds())
 	}()
 
+	statusUpdater := common.StatusUpdater{
```
Comment (wenyingd) [inline pkg/controllers/subnetset/subnetset_controller.go:58]: ditto

File: pkg/controllers/subnet/subnet_controller.go [hunk 7]
```diff
@@ -54,15 +54,26 @@ func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 	defer func() {
 		log.Info("Finished reconciling Subnet", "Subnet", req.NamespacedName, "duration(ms)", time.Since(startTime).Milliseconds())
 	}()
-	metrics.CounterInc(r.SubnetService.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnet)
 
+	statusUpdater := common.StatusUpdater{
```
Comment (wenyingd) [inline pkg/controllers/subnet/subnet_controller.go:58]: ditto

File: pkg/controllers/subnet/subnet_controller.go [hunk 8]
```diff
@@ -54,15 +54,26 @@ func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
+		Client:          r.Client,
+		NSXConfig:       r.SubnetService.NSXConfig,
+		Recorder:        r.Recorder,
+		MetricResType:   MetricResTypeSubnet,
+		NSXResourceType: "Subnet",
+		ResourceType:    "Subnet",
+	}
+
+	metrics.CounterInc(r.SubnetService.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnet)
```
Comment (wenyingd) [inline pkg/controllers/subnet/subnet_controller.go:67]: What do you feel if we also move the call of `metrics.CounterInc` to StatusUpdater ? If the effort is huge, we can update it in future with another change.

File: pkg/controllers/common/utils.go [hunk 9]
```diff
@@ -159,3 +162,53 @@ func GenericGarbageCollector(cancel chan bool, timeout time.Duration, f func(ctx
 		}
 	}
 }
+
+type UpdateSuccessStatusFn func(k8sclient.Client, context.Context, k8sclient.Object, metav1.Time)
+type UpdateFailStatusFn func(k8sclient.Client, context.Context, k8sclient.Object, metav1.Time, error)
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils.go]: Added, thanks

File: pkg/controllers/common/utils.go [hunk 10]
```diff
@@ -159,3 +162,53 @@ func GenericGarbageCollector(cancel chan bool, timeout time.Duration, f func(ctx
...
+	}
+	u.Recorder.Event(obj, v1.EventTypeNormal, ReasonSuccessfulUpdate, fmt.Sprintf("%s CR has been successfully updated", u.ResourceType))
+	metrics.CounterInc(u.NSXConfig, metrics.ControllerUpdateSuccessTotal, u.MetricResType)
+}
+
+func (u *StatusUpdater) UpdateFail(ctx context.Context, obj k8sclient.Object, err error, retry bool, msg string, setStatusFn UpdateFailStatusFn) {
+	if retry {
+		log.Error(err, fmt.Sprintf("Failed to create or update %s CR, would retry exponentially", u.ResourceType), "Reason", msg, u.ResourceType, obj)
+	} else {
+		log.Error(err, fmt.Sprintf("Failed to create or update %s CR", u.ResourceType), "Reason", msg, u.ResourceType, obj)
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils.go]: As discussed offline, I remove the retry part in this log and keep the log inside delete/update functions for common log sharing

File: pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go [hunk 11]
```diff
@@ -104,45 +84,53 @@ func (r *IPAddressAllocationReconciler) setReadyStatusTrue(ctx context.Context,
 func (r *IPAddressAllocationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	obj := &v1alpha1.IPAddressAllocation{}
 	log.Info("reconciling IPAddressAllocation CR", "IPAddressAllocation", req.NamespacedName)
+	statusUpdater := common.StatusUpdater{
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/ipaddressallocation/ipaddressallocation_controller.go]: Updated, thanks!

File: pkg/controllers/networkinfo/networkinfo_controller.go [hunk 12]
```diff
@@ -131,15 +131,26 @@ func (r *NetworkInfoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	defer func() {
 		log.Info("Finished reconciling NetworkInfo", "NetworkInfo", req.NamespacedName, "duration(ms)", time.Since(startTime).Milliseconds())
 	}()
+
+	statusUpdater := common.StatusUpdater{
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/networkinfo/networkinfo_controller.go]: Updated, and same below, thanks!

File: pkg/controllers/networkinfo/networkinfo_utils.go [hunk 13]
```diff
@@ -15,32 +14,17 @@ import (
...
 	svccommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
 )
 
-func deleteFail(r *NetworkInfoReconciler, c context.Context, o *v1alpha1.NetworkInfo, e *error, client client.Client) {
-	setNetworkInfoVPCStatus(c, o, client, nil)
-	r.Recorder.Event(o, v1.EventTypeWarning, common.ReasonFailDelete, fmt.Sprintf("%v", *e))
-	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, common.MetricResTypeNetworkInfo)
+func updateFail(c context.Context, o *v1alpha1.NetworkInfo, e error, vpcState *v1alpha1.VPCState, msg string, statusUpdater common.StatusUpdater) {
+	statusUpdater.UpdateFail(c, o, e, true, msg, nil)
+	setNetworkInfoVPCStatus(c, o, statusUpdater.Client, vpcState)
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/networkinfo/networkinfo_utils.go]: Updated, thanks!

File: pkg/controllers/securitypolicy/securitypolicy_controller.go [hunk 14]
```diff
@@ -65,27 +65,21 @@ func k8sClient(mgr ctrl.Manager) client.Client {
...
+	statusUpdater.DeleteFail(namespacedName, o, e)
+	if o != nil {
+		r.setSecurityPolicyReadyStatusFalse(c, o, metav1.Now(), e)
+	}
 }
 
-func updateSuccess(r *SecurityPolicyReconciler, c context.Context, o *v1alpha1.SecurityPolicy) {
+func updateSuccess(r *SecurityPolicyReconciler, c context.Context, o *v1alpha1.SecurityPolicy, statusUpdater common.StatusUpdater) {
+	statusUpdater.UpdateSuccess(c, o, nil)
 	r.setSecurityPolicyReadyStatusTrue(c, o, metav1.Now())
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/securitypolicy/securitypolicy_controller.go]: Updated, thanks

File: pkg/controllers/subnet/subnet_controller.go (hunk 8 above)
Comment (yanjunz97) [reply pkg/controllers/subnet/subnet_controller.go]: It's fine to include in this PR, updated, thanks!

File: pkg/controllers/common/utils_test.go [hunk 15]
```diff
@@ -226,3 +230,50 @@ func TestGetDefaultSubnetSet(t *testing.T) {
 	}
 
 }
+
+type fakeRecorder struct {
+}
+
+func (recorder fakeRecorder) Event(object runtime.Object, eventtype, reason, message string) {
+}
+func (recorder fakeRecorder) Eventf(object runtime.Object, eventtype, reason, messageFmt string, args ...interface{}) {
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils_test.go:240]: updated, thanks

**Review by wenyingd — COMMENTED** [review]
Overall LGTM, minor comments are about test.

File: pkg/controllers/common/utils_test.go [hunk 16]
```diff
@@ -226,3 +230,77 @@ func TestGetDefaultSubnetSet(t *testing.T) {
...
+}
+
+func TestStatusUpdater_UpdateFail(t *testing.T) {
+	statusUpdater := createStatusUpdater(t)
+	statusUpdater.UpdateFail(context.TODO(), &v1alpha1.Subnet{}, fmt.Errorf("mock error"), "log message", func(_ client.Client, _ context.Context, _ client.Object, _ metav1.Time, e error, _ ...interface{}) {
+		assert.Contains(t, e.Error(), "mock error")
+	})
+}
+
+func TestStatusUpdater_DeleteSuccess(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils_test.go:275]: I updated the test to add some verification to recorder calls, thanks

**Review by wenyingd — DISMISSED** [review]
LGTM

**Review by wenyingd — APPROVED** [review]
LGTM

**Review by zhengxiexie — APPROVED** [review]
LGTM

**Pull request merged**